tensorflow/compiler/mlir/tosa/g3doc/legalization.md

```diff
@@ -0,0 +1,4144 @@
+
+        // Step 4. Reshape transposed tensor to
+        // [ batch * prod(block_shape)] +
+        // [ padded_shape[1] / block_shape[0],
+        //   ...,
+        // padded_shape[M] / block_shape[M-1]] +
+        // remaining_shape
+
+        vector <size_t> a4_shape(%input.rank)
+        a4_shape[0] = batch_size * block_num_elements
+
+        for (int i = 0; i < block_rank; i++) {
+            a4_shape[i + 1] = %a1_pad.shape[i + 1] / %block.as_constant()[i]
+        }
+
+        for (int i = 0; i < remaining_block_shape; i++) {
+            a4_shape[1 + block_rank + i] = %input.shape[1 + block_rank + i]
+        }
+
+        %output = tosa.RESHAPE(%a3_reshape) {new_shape=a4_shape} : (tensor<%a3_reshape.type>) -> tensor<%output.type>
+
+        return %output
+    }
+
+lower_space_to_depth_op()
+-----------------------------
+
+    Value lower_space_to_depth_op(Value %input, size_t block_size[], Format_t data_format)
+    {
+        assert(data_format == 'NHWC')
+
+        vector <size_t> a2_shape = {%input.shape[0],
+                                    %input.shape[1] / block_size[0],
+                                    %block_size[0],
+                                    %input_shape[2] / block_size[1],
+                                    %block_size[1],
+                                    %input_shape[3]}
+        %a2_reshape = tosa.RESHAPE(%input) {new_shape=a2_shape} : (tensor<%input.type>) -> tensor<a2_shape, %input.dtype>
+        %a3_transpose = tosa.TRANSPOSE(%a2_reshape) {perm={0, 1, 3, 2, 4, 5}} : (tensor<%a2_reshape.type>) -> tensor<%a3_transpose.type>
+
+        vector <size_t> a4_shape = {%input.shape[0],
+                                    %input_shape[1] / block_size[0],
+                                    %input_shape[2] / block_size[1],
+                                    %input_shape[3] * block_size[0] * block_size[1]}
+        %output = tosa.RESHAPE(%a3_transpose) {new_shape=%a4_shape} : (tensor<%a3_transpose.type>) -> tensor<a4_shape, %input.dtype>
+        return %output
+    }
+
+lower_split_op()
+------------------
+
+    Value lower_split_op(Value %value, size_t axis, size_t num_split)
+    {
+        Value %output[]
+
+        size_t slice_size = %value.shape[axis] / num_split
+
+        for (int i = 0; i < num_split; i++) {
+            vector <size_t> begin_vals, size_vals
+
+            for (int j = 0; j < %value.rank; j++) {
+                if (j == axis) {
+                   begin_vals.push_back(slice_size * i)
+                   size_vals.push_back(slice_size)
+                } else {
+                   begin_vals.push_back(0)
+                   size_vals.push_bac(%value.shape[j])
+                }
+
+                %output[i] = tosa.SLICE(%value) {start=begin_vals, size=size_vals} (tensor<%value.type>) -> tensor<size_vals, %value.dtype>
+            }
+
+        }
+
+        %output_list = tosa.IDENTITYN(%output) (tensor<%output:*.type>) -> tensor<%output_list:*.type>
```

Comment:
We're all in agreement that this feels "wrong".  

Resolving this may be a simple matter of finding the right C++ syntax to form a list of operators to return from the legalization function - e.g., convertSplitOp() in legalize_common.cc, but we couldn't find an example of expressing this without introducing IDENTITYN.  Lowerings of Split, SplitV, and Unpack are situations where a list of results is needed and we couldn't find an alternative way to implement the lowering without introducing these seemingly-trivial ops.    

The question of "why does TOSA have IDENTITYN (or IDENTITY)?" came up in earlier discussions on review of the TOSA Dialect in LLVM.